File: finalists/César/challenge_19/19.java
import sys, base64

class BitStream:
    def __init__(self):
        self.bits = []

    def add(self, n, b=32):
        mask = 1 << (b - 1)
        while mask != 0:
            self.bits.append((n & mask) != 0)
            mask >>= 1

    def as_hex_string(self):
        padded_bits = self.bits + [0] * (7 - (len(self.bits) - 1) % 8)
        return "".join([self._convert_nibble(*padded_bits[i:i+4]) for i in range(0, len(padded_bits), 4)])

    def _convert_nibble(self, b3, b2, b1, b0):
        return "0123456789abcdef"[b3*8 + b2*4 + b1*2 + b0]

if __name__ == "__main__":
    # Extract the 32-bit elements from the input...
    text = sys.stdin.readline().strip()
    decoded_text = base64.b64decode(text)
    elements = [int(decoded_text[i:i+8], 16) for i in range(0, len(decoded_text), 8)]
    # ... encode into the bit stream...
    last = None
    bits = BitStream()
    for e in elements:
        if last != None and last - 16 <= e <= last + 15:
            bits.add(0, 1)
            bits.add(e - last, 5)
        else:
            bits.add(1, 1)
            bits.add(e)
        last = e
    # ... and print in base64.
    print base64.b64encode(bits.as_hex_string()).rstrip('=')

File: finalists/Daniel/challenge_19/19.py
#!/usr/bin/python

import sys, multiprocessing
import base64

def main(input):
	
	input = base64.b64decode(input)
	# Ok, let's start. First, put the input into an array of 32 bit integers
	alphabet = "0123456789abcdef"
	n_input = []
	counter = 7
	acc = 0
	for ch in input:
		acc <<= 4
		acc |= alphabet.index(ch)
		if counter == 0:
			n_input.append(acc)
			acc = 0
			counter = 7
		else:
			counter -= 1
	# Output will be a bit array
	output = []
	previous_value = 0
	# iterate through the input integers
	for i in n_input:
		# Compute the difference with the previous integer
		diff = i - previous_value
		if -0x10 <= diff <= 0xf: # if diff can be represented with 5 bits (sign included)
			# store a "0", then the diff in 5 bits
			output.append(0)
			# convert to positive
			if diff < 0:
				diff += 0x10
				diff |= 0x10
			bit = 5
			while bit:
				bit -= 1
				output.append((diff & (1 << bit)) >> bit)
		else:
			# store a "1", then the full integer value
			output.append(1)
			bit = 32
			while bit:
				bit -= 1
				output.append((i & (1 << bit)) >> bit)
		previous_value = i
	output.append(0)
	
	# now, compress the bit array into an hexadecimal one
	hex_output = []
	counter = 3
	acc = 0
	for b in output:
		acc <<= 1
		acc |= b
		if counter == 0:
			hex_output.append(alphabet[acc])
			acc = 0
			counter = 3
		else:
			counter -= 1
	if counter != 3:
		hex_output.append(alphabet[acc])
	
	# aaaand, last but not least, encode it in base64
	encoded_output = base64.b64encode(''.join(hex_output))
	# strip the padding ("==")
	encoded_output = encoded_output.split("=")[0]
	
	return encoded_output


if __name__ == '__main__':
	print(main(sys.stdin.read()))
